File: rust/dd-app/src/scanner.rs
//! Folder walking and the two-pass scan, ported from Scanner.swift.

use crate::decode;
use dd_core::fingerprint::{FingerprintEngine, FINGERPRINT_RATE};
use dd_core::hash::content_hash;
use dd_core::matcher;
use dd_core::model::{AudioFile, DuplicateGroup, MatchKind};
use dd_core::{formats, fingerprint::Fingerprint};
use rayon::prelude::*;
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use walkdir::WalkDir;

#[derive(Clone, Debug)]
pub struct ScanOptions {
    pub include_subfolders: bool,
    pub find_exact_duplicates: bool,
    pub find_similar_audio: bool,
    pub level: matcher::MatchLevel,
    pub minimum_duration: f64,
    pub fingerprint_seconds: f64,
    pub skip_hidden_files: bool,
}

impl Default for ScanOptions {
    fn default() -> Self {
        ScanOptions {
            include_subfolders: true,
            find_exact_duplicates: true,
            find_similar_audio: true,
            level: matcher::MatchLevel::Perfect,
            minimum_duration: 2.0,
            fingerprint_seconds: 120.0,
            skip_hidden_files: true,
        }
    }
}

#[derive(Default)]
pub struct ScanResult {
    pub groups: Vec<DuplicateGroup>,
    pub files_scanned: usize,
    pub files_skipped: Vec<(PathBuf, String)>,
}

impl ScanResult {
    pub fn reclaimable_bytes(&self) -> u64 {
        self.groups.iter().map(|g| g.reclaimable_bytes()).sum()
    }
}

#[derive(Clone, Copy, Debug)]
pub enum Phase {
    Collecting,
    Hashing { done: usize, total: usize },
    Fingerprinting { done: usize, total: usize },
    Matching,
    Finished,
}

/// Hidden-file rules differ by platform: a leading dot on Unix, the hidden
/// attribute on Windows.
#[cfg(windows)]
fn is_hidden(path: &Path) -> bool {
    use std::os::windows::fs::MetadataExt;
    const FILE_ATTRIBUTE_HIDDEN: u32 = 0x2;
    path.file_name()
        .and_then(|n| n.to_str())
        .map(|n| n.starts_with('.'))
        .unwrap_or(false)
        || std::fs::metadata(path)
            .map(|m| m.file_attributes() & FILE_ATTRIBUTE_HIDDEN != 0)
            .unwrap_or(false)
}

#[cfg(not(windows))]
fn is_hidden(path: &Path) -> bool {
    path.file_name().and_then(|n| n.to_str()).map(|n| n.starts_with('.')).unwrap_or(false)
}

/// A stable identity for a file, so the same one reached by two paths is only
/// counted once. Windows paths are case-insensitive.
fn canonical_identity(path: &Path) -> String {
    let resolved = std::fs::canonicalize(path).unwrap_or_else(|_| path.to_path_buf());
    let s = resolved.display().to_string();
    if cfg!(windows) { s.to_lowercase() } else { s }
}

pub fn collect(roots: &[PathBuf], options: &ScanOptions) -> Vec<AudioFile> {
    let mut seen = std::collections::HashSet::new();
    let mut files = Vec::new();

    for root in roots {
        let walker = WalkDir::new(root)
            .follow_links(false)
            .max_depth(if options.include_subfolders { usize::MAX } else { 1 });
        for entry in walker.into_iter().filter_map(|e| e.ok()) {
            let path = entry.path();
            if !entry.file_type().is_file() || !formats::is_audio(path) {
                continue;
            }
            if options.skip_hidden_files
                && path.components().any(|c| is_hidden(Path::new(c.as_os_str())))
            {
                continue;
            }
            let Ok(meta) = entry.metadata() else { continue };
            if meta.len() == 0 {
                continue;
            }
            if !seen.insert(canonical_identity(path)) {
                continue;
            }
            files.push(AudioFile::new(
                path.to_path_buf(),
                meta.len(),
                meta.modified().unwrap_or(std::time::SystemTime::UNIX_EPOCH),
            ));
        }
    }
    files.sort_by(|a, b| a.path.cmp(&b.path));
    files
}

pub fn run<F: Fn(Phase) + Sync>(
    roots: &[PathBuf],
    options: &ScanOptions,
    progress: F,
) -> ScanResult {
    let mut result = ScanResult::default();
    progress(Phase::Collecting);
    let mut files = collect(roots, options);
    result.files_scanned = files.len();
    if files.is_empty() {
        return result;
    }

    // Representative index -> every file byte-identical to it.
    let mut exact_members: HashMap<usize, Vec<usize>> = HashMap::new();
    let mut exact_groups: Vec<Vec<usize>> = Vec::new();

    // ---- Pass 1: byte-identical files -------------------------------
    if options.find_exact_duplicates {
        exact_groups = find_exact_duplicates(&mut files, &progress);
        for indices in &exact_groups {
            if let Some(&first) = indices.first() {
                exact_members.insert(first, indices.clone());
            }
        }
    }

    let mut similar_groups: Vec<(Vec<usize>, f64)> = Vec::new();

    // ---- Pass 2: same audio, different bytes ------------------------
    if options.find_similar_audio {
        // Hashing already proved the members of an exact group identical, so
        // only one of each needs to be decoded and fingerprinted.
        let collapsed: std::collections::HashSet<usize> = exact_members
            .values()
            .flat_map(|v| v.iter().cloned())
            .filter(|i| !exact_members.contains_key(i))
            .collect();
        let representatives: Vec<usize> =
            (0..files.len()).filter(|i| !collapsed.contains(i)).collect();

        let fingerprints = compute_fingerprints(&representatives, &mut files, options, &mut result, &progress);

        // Byte-identical files share the representative's audio properties.
        let pairs: Vec<(usize, Vec<usize>)> =
            exact_members.iter().map(|(k, v)| (*k, v.clone())).collect();
        for (representative, members) in pairs {
            for member in members {
                if member == representative {
                    continue;
                }
                files[member].duration = files[representative].duration;
                files[member].content_duration = files[representative].content_duration;
                files[member].sample_rate = files[representative].sample_rate;
                files[member].bit_depth = files[representative].bit_depth;
                files[member].channel_count = files[representative].channel_count;
                files[member].format_name = files[representative].format_name.clone();
                files[member].codec_is_lossless = files[representative].codec_is_lossless;
                files[member].fingerprint = files[representative].fingerprint.clone();
            }
        }

        progress(Phase::Matching);
        let durations: Vec<Option<f64>> = files.iter().map(|f| f.content_duration).collect();
        let matches = matcher::groups(
            &fingerprints,
            &durations,
            &matcher::Options::for_level(options.level),
        );
        similar_groups = matches
            .into_iter()
            .map(|g| {
                let members: Vec<usize> = g
                    .indices
                    .iter()
                    .flat_map(|i| exact_members.get(i).cloned().unwrap_or_else(|| vec![*i]))
                    .collect();
                (members, g.confidence)
            })
            .collect();
    }

    // An exact group wholly contained in a similar group is already represented
    // there; keeping both would double-count the wasted space.
    let absorbed: Vec<std::collections::HashSet<usize>> = similar_groups
        .iter()
        .map(|(m, _)| m.iter().cloned().collect())
        .collect();
    for indices in &exact_groups {
        let set: std::collections::HashSet<usize> = indices.iter().cloned().collect();
        if absorbed.iter().any(|a| a.is_superset(&set)) {
            continue;
        }
        result.groups.push(DuplicateGroup {
            kind: MatchKind::Exact,
            files: indices.iter().map(|&i| files[i].clone()).collect(),
            confidence: 1.0,
        });
    }
    for (members, confidence) in similar_groups {
        let mut sorted = members;
        sorted.sort_unstable();
        result.groups.push(DuplicateGroup {
            kind: MatchKind::Similar,
            files: sorted.iter().map(|&i| files[i].clone()).collect(),
            confidence,
        });
    }

    result.groups.sort_by(|a, b| b.reclaimable_bytes().cmp(&a.reclaimable_bytes()));
    progress(Phase::Finished);
    result
}

fn find_exact_duplicates<F: Fn(Phase) + Sync>(
    files: &mut [AudioFile],
    progress: &F,
) -> Vec<Vec<usize>> {
    // Only files sharing a byte size can possibly be identical.
    let mut by_size: HashMap<u64, Vec<usize>> = HashMap::new();
    for (i, f) in files.iter().enumerate() {
        by_size.entry(f.byte_size).or_default().push(i);
    }
    let mut candidates: Vec<usize> =
        by_size.values().filter(|v| v.len() > 1).flatten().cloned().collect();
    candidates.sort_unstable();
    if candidates.is_empty() {
        return Vec::new();
    }

    let total = candidates.len();
    progress(Phase::Hashing { done: 0, total });
    let done = std::sync::atomic::AtomicUsize::new(0);

    let hashes: Vec<(usize, String)> = candidates
        .par_iter()
        .filter_map(|&i| {
            let h = content_hash(&files[i].path).ok().map(|h| (i, h));
            let n = done.fetch_add(1, std::sync::atomic::Ordering::Relaxed) + 1;
            if n % 8 == 0 || n == total {
                progress(Phase::Hashing { done: n, total });
            }
            h
        })
        .collect();

    let mut by_hash: HashMap<String, Vec<usize>> = HashMap::new();
    for (i, h) in hashes {
        files[i].content_hash = Some(h.clone());
        by_hash.entry(h).or_default().push(i);
    }
    let mut groups: Vec<Vec<usize>> = by_hash
        .into_values()
        .filter(|v| v.len() > 1)
        .map(|mut v| {
            v.sort_unstable();
            v
        })
        .collect();
    groups.sort();
    groups
}

fn compute_fingerprints<F: Fn(Phase) + Sync>(
    indices: &[usize],
    files: &mut [AudioFile],
    options: &ScanOptions,
    result: &mut ScanResult,
    progress: &F,
) -> Vec<Option<Fingerprint>> {
    let total = indices.len();
    let mut fingerprints: Vec<Option<Fingerprint>> = vec![None; files.len()];
    if total == 0 {
        return fingerprints;
    }
    progress(Phase::Fingerprinting { done: 0, total });
    let done = std::sync::atomic::AtomicUsize::new(0);

    let paths: Vec<(usize, PathBuf)> = indices.iter().map(|&i| (i, files[i].path.clone())).collect();

    enum Outcome {
        Ok(usize, Box<decode::DecodedAudio>, dd_core::fingerprint::AudioAnalysis),
        Skip(PathBuf, String),
    }

    let outcomes: Vec<Outcome> = paths
        .par_iter()
        .map_init(FingerprintEngine::new, |engine, (index, path)| {
            let outcome = match decode::decode_for_fingerprint(path, options.fingerprint_seconds) {
                Ok(audio) => {
                    if audio.source_duration >= options.minimum_duration {
                        let analysis = engine.analyze(&audio.samples, FINGERPRINT_RATE);
                        Outcome::Ok(*index, Box::new(audio), analysis)
                    } else {
                        Outcome::Skip(
                            path.clone(),
                            format!("shorter than {}s", options.minimum_duration as i64),
                        )
                    }
                }
                Err(e) => Outcome::Skip(path.clone(), describe(&e, path)),
            };
            let n = done.fetch_add(1, std::sync::atomic::Ordering::Relaxed) + 1;
            if n % 4 == 0 || n == total {
                progress(Phase::Fingerprinting { done: n, total });
            }
            outcome
        })
        .collect();

    for outcome in outcomes {
        match outcome {
            Outcome::Ok(index, audio, analysis) => {
                fingerprints[index] = Some(analysis.fingerprint.clone());
                files[index].fingerprint = Some(analysis.fingerprint);
                files[index].duration = Some(audio.source_duration);
                files[index].sample_rate = Some(audio.source_sample_rate);
                files[index].bit_depth = audio.source_bit_depth;
                files[index].channel_count = Some(audio.source_channels);
                files[index].format_name = Some(audio.format_name.clone());
                files[index].codec_is_lossless = Some(audio.is_lossless);

                // Padding is not content. When the decode stopped early the
                // tail was never seen, so only the head silence is discounted.
                let trailing = if audio.was_truncated { 0.0 } else { analysis.trailing_silence };
                files[index].content_duration =
                    Some((audio.source_duration - analysis.leading_silence - trailing).max(0.0));
            }
            Outcome::Skip(path, reason) => result.files_skipped.push((path, reason)),
        }
    }
    fingerprints
}

fn describe(error: &decode::DecodeError, path: &Path) -> String {
    if formats::is_byte_comparable_only(path) {
        return format!(
            "no {} decoder available — only exact copies of this file can be found",
            formats::label(path)
        );
    }
    error.to_string()
}

File: Sources/DupliDetect/Scanner.swift
import Foundation

public struct ScanOptions {
    public var includeSubfolders = true
    public var findExactDuplicates = true
    public var findSimilarAudio = true
    public var matchLevel: FingerprintMatcher.Level = .perfect
    /// Ignore files shorter than this; very short clips match each other easily.
    public var minimumDuration: Double = 2.0
    /// Only fingerprint the first N seconds of each file.
    public var fingerprintSeconds: Double = 120
    public var skipHiddenFiles = true

    public init() {}
}

public enum ScanPhase: Equatable {
    case idle
    case collecting
    case hashing(done: Int, total: Int)
    case fingerprinting(done: Int, total: Int)
    case matching
    case finished

    public var describes: String {
        switch self {
        case .idle: return ""
        case .collecting: return "Looking for audio files…"
        case .hashing(let done, let total): return "Checking for identical files… \(done) of \(total)"
        case .fingerprinting(let done, let total): return "Listening to audio… \(done) of \(total)"
        case .matching: return "Comparing fingerprints…"
        case .finished: return "Done"
        }
    }

    public var fraction: Double? {
        switch self {
        case .hashing(let done, let total),
             .fingerprinting(let done, let total):
            return total > 0 ? Double(done) / Double(total) : nil
        default: return nil
        }
    }
}

public struct ScanResult {
    public var groups: [DuplicateGroup] = []
    public var filesScanned = 0
    public var filesSkipped: [(url: URL, reason: String)] = []
    /// The level this scan actually ran at, so the results keep describing
    /// themselves correctly even if the setting is changed afterwards.
    public var matchLevel: FingerprintMatcher.Level = .perfect

    public var reclaimableBytes: Int64 { groups.reduce(0) { $0 + $1.reclaimableBytes } }
}

public final class Scanner {

    private let cancelled = Atomic(false)

    public init() {}

    public func cancel() { cancelled.value = true }

    public func run(roots: [URL],
                    options: ScanOptions,
                    progress: @escaping (ScanPhase) -> Void) -> ScanResult {
        cancelled.value = false
        var result = ScanResult()
        result.matchLevel = options.matchLevel

        progress(.collecting)
        var files = collect(roots: roots, options: options)
        result.filesScanned = files.count
        if cancelled.value || files.isEmpty { return result }

        // Groups are accumulated as index lists and only turned into
        // DuplicateGroups at the very end, once every file carries its metadata.
        var exactGroups: [[Int]] = []
        var similarGroups: [(indices: [Int], confidence: Double)] = []

        // Representative index -> every file byte-identical to it.
        var exactMembers: [Int: [Int]] = [:]

        // ---- Pass 1: byte-identical files -------------------------------
        if options.findExactDuplicates {
            exactGroups = findExactDuplicates(in: &files, progress: progress)
            for indices in exactGroups {
                guard let representative = indices.first else { continue }
                exactMembers[representative] = indices
            }
        }
        if cancelled.value { return result }

        // ---- Pass 2: same audio, different bytes ------------------------
        if options.findSimilarAudio {
            // Hashing already proved the members of an exact group identical,
            // so only one of each needs to be decoded and fingerprinted.
            let collapsed = Set(exactMembers.values.flatMap { $0 }).subtracting(exactMembers.keys)
            let representatives = files.indices.filter { !collapsed.contains($0) }

            let fingerprints = computeFingerprints(for: representatives,
                                                   files: &files,
                                                   options: options,
                                                   result: &result,
                                                   progress: progress)
            if cancelled.value { return result }

            // Byte-identical files share the representative's audio properties.
            for (representative, members) in exactMembers {
                for member in members where member != representative {
                    files[member].duration = files[representative].duration
                    files[member].contentDuration = files[representative].contentDuration
                    files[member].sampleRate = files[representative].sampleRate
                    files[member].bitDepth = files[representative].bitDepth
                    files[member].codecIsLossless = files[representative].codecIsLossless
                    files[member].channelCount = files[representative].channelCount
                    files[member].formatName = files[representative].formatName
                    files[member].fingerprint = files[representative].fingerprint
                }
            }

            progress(.matching)
            let matcherOptions = FingerprintMatcher.Options.forLevel(options.matchLevel)
            let matches = FingerprintMatcher.groups(fingerprints: fingerprints,
                                                    durations: files.map(\.contentDuration),
                                                    options: matcherOptions,
                                                    shouldCancel: { self.cancelled.value })

            // Expand each representative back into its exact-duplicate members.
            similarGroups = matches.map { (indices: $0.indices.flatMap { exactMembers[$0] ?? [$0] },
                                           confidence: $0.confidence) }
        }

        // An exact group wholly contained in a similar group is already
        // represented there; keeping both would double-count the wasted space.
        let absorbed = similarGroups.map { Set($0.indices) }
        for indices in exactGroups where !absorbed.contains(where: { $0.isSuperset(of: indices) }) {
            result.groups.append(DuplicateGroup(kind: .exact,
                                                files: indices.map { files[$0] },
                                                confidence: 1.0))
        }
        for group in similarGroups {
            result.groups.append(DuplicateGroup(kind: .similar,
                                                files: group.indices.sorted().map { files[$0] },
                                                confidence: group.confidence))
        }

        result.groups.sort { $0.reclaimableBytes > $1.reclaimableBytes }
        progress(.finished)
        return result
    }

    // MARK: - Collecting

    private func collect(roots: [URL], options: ScanOptions) -> [AudioFile] {
        var seen = Set<String>()
        var files: [AudioFile] = []
        let keys: [URLResourceKey] = [.isRegularFileKey, .fileSizeKey, .contentModificationDateKey, .isSymbolicLinkKey]

        for root in roots {
            if cancelled.value { break }
            var isDirectory: ObjCBool = false
            guard FileManager.default.fileExists(atPath: root.path, isDirectory: &isDirectory) else { continue }

            if !isDirectory.boolValue {
                append(root, to: &files, seen: &seen, keys: keys)
                continue
            }

            var enumeratorOptions: FileManager.DirectoryEnumerationOptions = [.skipsPackageDescendants]
            if options.skipHiddenFiles { enumeratorOptions.insert(.skipsHiddenFiles) }
            if !options.includeSubfolders { enumeratorOptions.insert(.skipsSubdirectoryDescendants) }

            guard let enumerator = FileManager.default.enumerator(at: root,
                                                                  includingPropertiesForKeys: keys,
                                                                  options: enumeratorOptions) else { continue }
            for case let url as URL in enumerator {
                if cancelled.value { break }
                append(url, to: &files, seen: &seen, keys: keys)
            }
        }
        return files
    }

    private func append(_ url: URL, to files: inout [AudioFile], seen: inout Set<String>, keys: [URLResourceKey]) {
        guard AudioFormats.isAudio(url) else { return }
        guard let values = try? url.resourceValues(forKeys: Set(keys)),
              values.isRegularFile == true,
              let size = values.fileSize, size > 0 else { return }
        // Resolve so the same file reached by two paths is only counted once.
        let identity = url.resolvingSymlinksInPath().standardizedFileURL.path
        guard seen.insert(identity).inserted else { return }
        var file = AudioFile(url: url,
                             byteSize: Int64(size),
                             modified: values.contentModificationDate ?? .distantPast)
        // A sensible name to show even if the file is never decoded.
        file.formatName = AudioFormats.label(for: url)
        files.append(file)
    }

    // MARK: - Pass 1

    private func findExactDuplicates(in files: inout [AudioFile],
                                     progress: @escaping (ScanPhase) -> Void) -> [[Int]] {
        // Only files sharing a byte size can possibly be identical, so hash nothing else.
        var bySize: [Int64: [Int]] = [:]
        for index in files.indices { bySize[files[index].byteSize, default: []].append(index) }
        let candidates = bySize.values.filter { $0.count > 1 }.flatMap { $0 }.sorted()
        guard !candidates.isEmpty else { return [] }

        // Snapshot the URLs; the worker closures must not touch `files` itself.
        let urls = candidates.map { files[$0].url }
        let total = candidates.count
        progress(.hashing(done: 0, total: total))

        let done = Atomic(0)
        let hashes = Atomic([Int: String]())
        let group = DispatchGroup()
        let queue = DispatchQueue(label: "duplidetect.hash", attributes: .concurrent)
        let workerCount = max(1, min(total, ProcessInfo.processInfo.activeProcessorCount))

        for worker in 0..<workerCount {
            queue.async(group: group) {
                var slot = worker
                while slot < total {
                    defer { slot += workerCount }
                    if self.cancelled.value { return }
                    if let hash = try? FileHasher.contentHash(of: urls[slot]) {
                        hashes.mutate { $0[candidates[slot]] = hash }
                    }
                    let count = self.increment(done)
                    if count % 8 == 0 || count == total {
                        progress(.hashing(done: count, total: total))
                    }
                }
            }
        }
        group.wait()
        if cancelled.value { return [] }

        let resolved = hashes.value
        for (index, hash) in resolved { files[index].contentHash = hash }

        var byHash: [String: [Int]] = [:]
        for (index, hash) in resolved { byHash[hash, default: []].append(index) }
        return byHash.values.filter { $0.count > 1 }.map { $0.sorted() }
    }

    private func increment(_ counter: Atomic<Int>) -> Int {
        var value = 0
        counter.mutate { $0 += 1; value = $0 }
        return value
    }

    // MARK: - Pass 2

    private func computeFingerprints(for indices: [Int],
                                     files: inout [AudioFile],
                                     options: ScanOptions,
                                     result: inout ScanResult,
                                     progress: @escaping (ScanPhase) -> Void) -> [Fingerprint?] {

        var fingerprints = [Fingerprint?](repeating: nil, count: files.count)
        let total = indices.count
        guard total > 0 else { return fingerprints }
        progress(.fingerprinting(done: 0, total: total))

        let done = Atomic(0)
        let collected = Atomic([Int: (AudioAnalysis, DecodedAudio)]())
        let skipped = Atomic([(URL, String)]())

        let workerCount = max(1, min(indices.count, ProcessInfo.processInfo.activeProcessorCount))
        let queue = DispatchQueue(label: "duplidetect.fingerprint", attributes: .concurrent)
        let group = DispatchGroup()
        let urls = indices.map { files[$0].url }

        for worker in 0..<workerCount {
            queue.async(group: group) {
                // One FFT setup per worker; the setup owns scratch buffers.
                let engine = FingerprintEngine()
                var slot = worker
                while slot < total {
                    defer { slot += workerCount }
                    if self.cancelled.value { return }

                    let index = indices[slot]
                    do {
                        let audio = try AudioDecoder.decodeMono(url: urls[slot],
                                                                maxSeconds: options.fingerprintSeconds)
                        if audio.sourceDuration >= options.minimumDuration {
                            let analysis = engine.analyze(samples: audio.samples)
                            collected.mutate { $0[index] = (analysis, audio) }
                        } else {
                            skipped.mutate { $0.append((urls[slot], "shorter than \(Int(options.minimumDuration))s")) }
                        }
                    } catch {
                        skipped.mutate { $0.append((urls[slot], Self.describe(error, url: urls[slot]))) }
                    }

                    let count = self.increment(done)
                    if count % 4 == 0 || count == total {
                        progress(.fingerprinting(done: count, total: total))
                    }
                }
            }
        }
        group.wait()

        for (index, payload) in collected.value {
            let (analysis, audio) = payload
            fingerprints[index] = analysis.fingerprint
            files[index].fingerprint = analysis.fingerprint
            files[index].duration = audio.sourceDuration
            files[index].sampleRate = audio.sourceSampleRate
            files[index].bitDepth = audio.sourceBitDepth
            files[index].codecIsLossless = audio.isLossless
            files[index].channelCount = audio.sourceChannels
            files[index].formatName = audio.formatName

            // Padding is not content. When the decode stopped early the tail was
            // never seen, so only the head silence can be discounted.
            let trailing = audio.wasTruncated ? 0 : analysis.trailingSilence
            files[index].contentDuration = max(0, audio.sourceDuration - analysis.leadingSilence - trailing)
        }
        result.filesSkipped.append(contentsOf: skipped.value.map { (url: $0.0, reason: $0.1) })
        return fingerprints
    }

    private static func describe(_ error: Error, url: URL) -> String {
        if AudioFormats.byteComparableOnly.contains(url.pathExtension.lowercased()) {
            return "macOS has no \(AudioFormats.label(for: url)) decoder — only exact copies of this file can be found"
        }
        switch error {
        case AudioDecodeError.unsupported(let detail): return "could not be decoded (\(detail))"
        case AudioDecodeError.emptyAudio: return "contains no audio"
        case AudioDecodeError.decodeFailed(let detail): return "decode failed (\(detail))"
        default: return error.localizedDescription
        }
    }
}

/// Small lock-protected box; the scan runs across several queues.
public final class Atomic<Value> {
    private var storage: Value
    private let lock = NSLock()

    public init(_ value: Value) { storage = value }

    public var value: Value {
        get { lock.lock(); defer { lock.unlock() }; return storage }
        set { lock.lock(); storage = newValue; lock.unlock() }
    }

    public func mutate(_ body: (inout Value) -> Void) {
        lock.lock(); body(&storage); lock.unlock()
    }
}
